### PR TITLE
Devcenter environments Outputs API Integration

### PR DESCRIPTION
**Pull request by wbreza**
This update moves away from manually reading Azure deployment outputs to leveraging the Outputs API directly from the ADE team.  This enables a seamless outputs experience across their supported runners including ARM, Bicep, Terraform and other custom runners.

> [!IMPORTANT]
> There is a blocking bug on the ADE side around the serialization of `array` and `object` outputs that need to be resolved before we can move forward.

### CODE REVIEW

**Comment by ellismg** [comment]
> There is a blocking bug on the ADE side around the serialization of array and object outputs that need to be resolved before we can move forward.

@wbreza - I know you're going to be on vacation for a few weeks - is there anything you want us to keep track of with respect to the above while you're gone (or a sample project we should be testing with to see when the fix has landed) or do you just want to have this on hold while you are gone (looking at the diff, I suspect the chances of it bitrotting while you are out is small, but I want if there's something we can do while you're gone to keep things moving along, happy to do it).

File: cli/azd/pkg/devcenter/manager.go
Comment (ellismg) [inline cli/azd/pkg/devcenter/manager.go:377]: Just a note - this uppercasing was working around a very specific bug related to ARM deployments, where the casing of outputs was not preserved in the deployment object.  Now that this `createOutputParameters` thing applies to all forms of IaC is doing this casing still right?  Should we be nervous there?

I sort of wish the ADE folks could update their ARM runner to just do the casing normalization on their side, since they would have the original deployment template and could ensure the casing was right, and we just trust that stuff from the `Outputs` API is cased correctly.

**Review by ellismg — COMMENTED** [review]
Overall looks good - one question about the casing hack we now seem to apply more generally?  I would "Approve" except I'm not really sure if you want that yet due to the upstream ADE bug you need a fix for.

File: cli/azd/pkg/devcenter/manager.go
Comment (wbreza) [reply cli/azd/pkg/devcenter/manager.go:377]: The ADE team does not currently perform any normalization at their API level.  I could see both sides of this since they are just returning the same casing as you would see from the deployments API directly.  At the same time we know this is incorrect and is not honoring the casing set by the end user in the ARM/Bicep.

If the ARM team ever fixes this they could inadvertently normalize to an unexpected casing.  My 2 cents would be to leave it up to the final consuming apps/users like us??

**Comment by wbreza** [comment]
> > There is a blocking bug on the ADE side around the serialization of array and object outputs that need to be resolved before we can move forward.
> 
> @wbreza - I know you're going to be on vacation for a few weeks - is there anything you want us to keep track of with respect to the above while you're gone (or a sample project we should be testing with to see when the fix has landed) or do you just want to have this on hold while you are gone (looking at the diff, I suspect the chances of it bitrotting while you are out is small, but I want if there's something we can do while you're gone to keep things moving along, happy to do it).

I would just keep an heads up on the ADE bi-weekly sync and keep asking about this issue.  There isn't any extreme urgency for this to get in before I return.

File: cli/azd/pkg/devcenter/manager.go
Comment (weikanglim) [inline cli/azd/pkg/devcenter/manager.go:395]: Just checking on this -- what requires us to marshal complex types? It's interesting that we'd allow for `int` and `bool` to be set directly, but we'd represent the complex types as JSON-string. 

I know in [provisioning manager](https://github.com/Azure/azure-dev/blob/e03b5fccc87cceafce9e83f16688b3c7a7bb8a97/cli/azd/pkg/infra/provisioning/manager.go#L156), we'd end up marshaling it for environment outputting, but the Bicep provider doesn't serialize the output either, it just passes the complex types through.

File: cli/azd/pkg/devcenter/manager.go
Comment (weikanglim) [reply cli/azd/pkg/devcenter/manager.go:377]: The one thing that @ellismg mentions, that may be the thing we'd want to avoid:

>Now that this createOutputParameters thing **applies to all forms of IaC** is doing this casing still right? Should we be nervous there?

Comment (ellismg) [reply cli/azd/pkg/devcenter/manager.go:377]: > If the ARM team ever fixes this they could inadvertently normalize to an unexpected casing. My 2 cents would be to leave it up to the final consuming apps/users like us??

My hope is that one day the ARM team does fix the casing issues (I suspect they will have to gate it behind an `api-version` bump at a minimum, but maybe they will have to have some whole new API due to how they view breaking changes).

The original design (before we tried to support BYO scenarios) would use the source ARM template to recover the casing of the original outputs before presenting it to the rest of the system.  For BYO scenarios, where we didn't have the template, we just did this ToUpper which felt unsatisfying but was the best we could do given our conventions.

My hope would be that whatever ARM runner ADE uses would end up playing the same tricks we did (we should be possible, since they have the source template) to ensure that the outputs are correctly cased and then `azd` could just pass things as is. I'll mention this to the ADE folks.

As @weikanglim - my larger fear here is that if we have to do something special for ARM type deployments to work around this bug, we do it in a way that only applies to ADE environments backed by ARM Templates.  I am afraid that if we start `ToUpper`ing stuff from Terraform, Pulumi or other custom runners that don't have this bug we are going to end up breaking customer applications, which don't use our `ALL_UPPERCASE` convention for outputs.

Comment (wbreza) [reply cli/azd/pkg/devcenter/manager.go:377]: This whole normalization to UPPER isn't anything new.  We've been doing this for a very long time and this just continues this normalization process to during the usage of the outputs API instead of the raw ARM deployments API

File: cli/azd/pkg/devcenter/manager.go
Comment (wbreza) [reply cli/azd/pkg/devcenter/manager.go:395]: I believe that ARM deployments API already returns complex types as JSON strings and this is just matching that expectation since these outputs are being set in `.env` files.

File: cli/azd/pkg/devcenter/manager.go
Comment (weikanglim) [reply cli/azd/pkg/devcenter/manager.go:377]: Did we previously only do it for ARM and not Terraform? Do we want to avoid doing it for Terraform for the future?

File: cli/azd/pkg/devcenter/manager.go
Comment (weikanglim) [reply cli/azd/pkg/devcenter/manager.go:395]: Indeed the API is returning a JSON string.

When we call `json.Unmarshal` however, the value would be deserialized as a weakly-typed `map[string]interface{}`. I'd like to keep a consistent model of what `Outputs` should represent -- it's the deserialized JSON (weakly-typed). Serializing it here as a JSON string makes it harder to reason about. 

Keeping it as `map[string]interface{}` instead of `string` also make ADE's provision provider consistent with other providers. Here is what happens today in Bicep provisioning provider when it returns the `Outputs` to the provision manager, it's a `map[string]interface{}`:
image

We only serialize all values to their string-representation when we're saving it to the `.env` file in provision.Manager:
image

My best guess is that if we removed JSON marshaling here, the observable-behavior of the system would work just the same, we just made the internal representation more consistent.

File: cli/azd/pkg/devcenter/manager.go
Comment (wbreza) [reply cli/azd/pkg/devcenter/manager.go:377]: The IaC technology used by a Devcenter environment definition is an internal implementation detail of the template and `azd` doesn't know.  Previously Devcenter only supported ARM templates, but now is open ended and supports, ARM, Bicep & Terraform.

To that end I could see an issue that we are normalizing for all devcenter IaC providers instead of only ARM based runners.

File: cli/azd/pkg/devcenter/manager.go
Comment (wbreza) [reply cli/azd/pkg/devcenter/manager.go:395]: Ok, so you are saying that our `UpdateEnvironment` func already handles the complexities of marshalling complex types into string... If that is the case then I agree that the Outputs should be consistent with this experience.

**Review by weikanglim — APPROVED** [review]

**Pull request closed without merging**
